{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true,
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "# Pooling-imputation performance from real bead chip data with high density markers\n",
    "\n",
    "Performs pooling simulation and imputation on data from the chromosome 20 of 1000GP.\n",
    "Markers have been chosen as the intersection between the real bead chip Illumina Infinium OmniExpress2.5 - 8 Kit\n",
    " (https://support.illumina.com/array/array_kits/humanomniexpress2.5-8-beadchip-kit/downloads.html) \n",
    " and the chr20 1000GP data. The samples are randomly assigned to the reference panel or the study population.\n",
    "\n",
    "Apply pooling simulation on this data, and imputation (run parallel execution): \n",
    "**pool and impute bead chip markers only, compute metrics and plot statistics** \n",
    "Imputation has been with increased **modelscale=3.0** parameter for trying to account \n",
    "for dependency between clustered genotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "try:\n",
    "    os.mkdir('/home/camille/PoolImpHuman/data/20200717')\n",
    "except FileExistsError:\n",
    "    pass\n",
    "os.chdir('/home/camille/PoolImpHuman/data/20200717')"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n",
     "is_executing": false
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "outputs": [
    {
     "name": "stdout",
     "text": [
      "Configure directory\n"
     ],
     "output_type": "stream"
    },
    {
     "data": {
      "text/plain": "[\"ln: failed to create symbolic link './InfiniumOmniExpress-chr20-CHROM-POS.txt': File exists\"]"
     },
     "metadata": {},
     "output_type": "execute_result",
     "execution_count": 2
    }
   ],
   "source": [
    "print('Configure directory')\n",
    "%sx ln -s ~/1000Genomes/scripts/VCFPooling/python/omniexpress_20200717.ipynb ./\n",
    "%sx ln -s ../omniexpress8/InfiniumOmniExpress-chr20-CHROM-POS.txt ./"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n",
     "is_executing": false
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Prepare experimental VCF file\n",
    "\n",
    "IMP and REF files are identical to 20200709 (imputation with default parameters)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [
    {
     "name": "stdout",
     "text": [
      "Impute missing genotypes in the pooled file\n"
     ],
     "output_type": "stream"
    }
   ],
   "source": [
    "print('Impute missing genotypes in the pooled file')\n",
    "# done on Rackham"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n",
     "is_executing": false
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "outputs": [
    {
     "name": "stdout",
     "text": [
      "Plotting results with bcftools stats\n"
     ],
     "output_type": "stream"
    },
    {
     "data": {
      "text/plain": "['Parsing bcftools stats output: filestats.vchk',\n 'Sanity check failed: was this file generated by bcftools stats? at /usr/bin/plot-vcfstats line 99.',\n '\\tmain::error(\"Sanity check failed: was this file generated by bcftools stats?\") called at /usr/bin/plot-vcfstats line 585',\n '\\tmain::parse_vcfstats1(HASH(0x560e817f3b78), 0) called at /usr/bin/plot-vcfstats line 294',\n '\\tmain::parse_vcfstats(HASH(0x560e817f3b78)) called at /usr/bin/plot-vcfstats line 47']"
     },
     "metadata": {},
     "output_type": "execute_result",
     "execution_count": 4
    }
   ],
   "source": [
    "print('Plotting results with bcftools stats')\n",
    "%sx deactivate\n",
    "# bcftools stats needs python 2.7\n",
    "%sx bcftools stats --af-bins 0.01,0.02,0.04,0.08,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,0.98 --collapse snps -S study.population IMP.chr20.pooled.imputed.vcf.gz IMP.chr20.snps.gt.vcf.gz > filestats.vchk\n",
    "%sx plot-vcfstats -p bcftoolstats -s filestats.vchk"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n",
     "is_executing": false
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Compute results with customized metrics"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% md\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "outputs": [],
   "source": [
    "paths = {'beaglegt': {\n",
    "    'true': '/home/camille/PoolImpHuman/data/20200709/IMP.chr20.snps.gt.vcf.gz',\n",
    "    'imputed': '/home/camille/PoolImpHuman/data/20200709/IMP.chr20.pooled.imputed.vcf.gz'},\n",
    "         'beaglegl': {\n",
    "     'true': '/home/camille/PoolImpHuman/data/20200709/IMP.chr20.snps.gl.vcf.gz',\n",
    "     'imputed': '/home/camille/PoolImpHuman/data/20200709/IMP.chr20.pooled.imputed.vcf.gz'},\n",
    "}"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n",
     "is_executing": false
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "convertgtgl = True\n",
    "if convertgtgl:\n",
    "    cmd = 'bash ~/PoolImpHuman/bin/bash-scripts/gt_to_gl.sh {} {}'.format(paths['beaglegt']['true'], paths['beaglegl']['true'])\n",
    "    subprocess.run(cmd, shell=True,)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n",
     "is_executing": false
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "outputs": [
    {
     "name": "stdout",
     "text": [
      "['/home/camille/PoolImpHuman/data/20200717', '/home/camille/1000Genomes', '/usr/lib/python36.zip', '/usr/lib/python3.6', '/usr/lib/python3.6/lib-dynload', '', '/home/camille/1000Genomes/venv3.6/lib/python3.6/site-packages', '/home/camille/1000Genomes/venv3.6/lib/python3.6/site-packages/setuptools-40.8.0-py3.6.egg', '/home/camille/1000Genomes/venv3.6/lib/python3.6/site-packages/pip-19.0.3-py3.6.egg', '/home/camille/1000Genomes/venv3.6/lib/python3.6/site-packages/IPython/extensions', '/home/camille/.ipython']\n"
     ],
     "output_type": "stream"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from scripts.VCFPooling.poolSNPs.metrics import quality"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n",
     "is_executing": false
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "outputs": [],
   "source": [
    "qbeaglegt = quality.QualityGT(*paths['beaglegt'].values(), 0, idx='id')"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n",
     "is_executing": false
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "outputs": [],
   "source": [
    "qbeaglegl = quality.QualityGL(paths['beaglegl']['true'], paths['beaglegl']['imputed'], 0, idx='id')\n",
    "messbeagle = qbeaglegl.cross_entropy"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%% \n",
     "is_executing": false
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "#qbeaglegl = quality.QualityGT(*paths['beaglegt'].values(), 0, idx='id')\n",
    "tabbeaglegl = pd.concat([qbeaglegt.concordance(),\n",
    "                       qbeaglegt.trueobj.af_info,\n",
    "                       qbeaglegt.pearsoncorrelation(),\n",
    "                       qbeaglegt.precision,\n",
    "                       qbeaglegt.accuracy,\n",
    "                       qbeaglegt.recall,\n",
    "                       qbeaglegt.f1_score,\n",
    "                        qbeaglegl.cross_entropy], axis=1)\n",
    "dosbeaglegl = qbeaglegt.alleledosage()\n"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n",
     "is_executing": true
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "tabbeaglegl.head()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n",
     "is_executing": true
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "plt.rcParams[\"figure.figsize\"] = [5*4, 4*2]\n",
    "fig, axes = plt.subplots(2, 4)\n",
    "\n",
    "tabbeaglegl.plot.scatter('af_info', 'precision_score', ax=axes[0, 0], s=0.7, label='beaglegl')\n",
    "axes[0, 0].set_ylim(0.0, 1.0)\n",
    "tabbeaglegl.plot.scatter('af_info', 'accuracy_score', ax=axes[0, 1], s=0.7, label='beaglegl')\n",
    "axes[0, 1].set_ylim(0.0, 1.0)\n",
    "tabbeaglegl.plot.scatter('af_info', 'concordance', ax=axes[0, 2], s=0.7, label='beaglegl')\n",
    "axes[0, 2].set_ylim(0.0, 1.0)\n",
    "tabbeaglegl.plot.scatter('af_info', 'f1_score', ax=axes[0, 3], s=0.7, label='beaglegl')\n",
    "axes[0, 3].set_ylim(0.0, 1.0)\n",
    "tabbeaglegl.plot.scatter('af_info', 'r_squared', ax=axes[1, 0], s=0.7, label='beaglegl')\n",
    "axes[1, 0].set_ylim(-0.2, 1.0)\n",
    "tabbeaglegl.plot.scatter('af_info', 'cross_entropy', ax=axes[1, 1], s=0.7, label='beaglegl')\n",
    "axes[1, 1].set_ylim(-0.5, 5.0)\n",
    "axes[1, 2].scatter(dosbeaglegl[0], dosbeaglegl[1], s=0.7, label='beaglegl')\n",
    "axes[1, 2].set_xlabel('true allele dosage')\n",
    "axes[1, 2].set_ylabel('imputed allele dosage')\n",
    "axes[1, 2].set_ylim(0.0, 2.0)\n",
    "\n",
    "for ax in axes.flatten()[:-2]:\n",
    "    # cast color to white 'w' if dark background\n",
    "    ax.set_xlabel('true alternate allele frequency', color='w')\n",
    "    ax.set_ylabel(ax.get_ylabel(), color='w')\n",
    "plt.savefig(os.path.join(os.path.dirname(paths['beaglegt']['imputed']), 'imputation_quality_gtgl.png'))\n",
    "plt.show()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n",
     "is_executing": true
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  },
  "pycharm": {
   "stem_cell": {
    "cell_type": "raw",
    "source": [],
    "metadata": {
     "collapsed": false
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}